src/services/gameleaderboard.service.ts

```diff
@@ -0,0 +1,280 @@
+import { EmbedFieldData, MessageEmbed, Snowflake, User } from 'discord.js';
+import { Collection } from 'mongodb';
+import { Maybe } from '../common/types';
+import { GuildService } from './guild.service';
+import { LoggerService } from './logger.service';
+import { StorageService } from './storage.service';
+
+export class GameLeaderboardService {
+  public gameAliases: Record<GameType, string[]> = {
+    [GameType.TicTacToe]: ['ttt', 'tictactoe'],
+    [GameType.ConnectFour]: ['c4', 'connectfour', 'connect-four'],
+  };
+
+  private _gameEnumToString: Record<GameType, string> = {
+    [GameType.TicTacToe]: 'Tic-Tac-Toe',
+    [GameType.ConnectFour]: 'Connect 4',
+  };
+
+  private _gameEnumToCollection: Record<GameType, Function> = {
+    [GameType.TicTacToe]: this._getCollection('tttLeaderboard'),
+    [GameType.ConnectFour]: this._getCollection('connectFourLeaderboard'),
+  };
+
+  constructor(
+    private _storageService: StorageService,
+    private _guildService: GuildService,
+    private _loggerService: LoggerService
+  ) {}
+
+  public async updateLeaderboard(user: User, game: GameType, gameData: IGame) {
+    const leaderboard = await this._gameToCollection(game);
+
+    if (!leaderboard) {
+      this._loggerService.error(`Could not get leaderboard for ${game}`);
+      return;
+    }
+
+    // Get the entry for the user
+    let userDoc = await leaderboard.findOne({
+      userId: user.id,
+      guildId: this._guildService.get().id,
+    });
+
+    // if the user doesnt have an entry yet, make one for them
+    if (!userDoc) {
+      await leaderboard.insertOne({
+        userId: user.id,
+        guildId: this._guildService.get().id,
+        games: [],
+      });
+
+      userDoc = await leaderboard.findOne({
+        userId: user.id,
+        guildId: this._guildService.get().id,
+      });
+    }
+
+    if (!userDoc) {
+      this._loggerService.error(
+        `Failed to make or find entry for user with id ${user.id} in leaderboard ${leaderboard} for game ${game}`
+      );
+      return;
+    }
+
+    // update the user's document by pushing the new game data
+    // to the user's array
+    await leaderboard.updateOne(
+      { userId: user.id },
+      { $push: { games: { $each: [gameData], $position: 0 } } }
+    );
+  }
+
+  private _getCollection(
+    gameType: GameCollectionTypes
+  ): (game: GameType) => Promise<Maybe<Collection<IGameLeaderBoardEntry>>> {
+    const collections = this._storageService.getCollections();
+    return async () => (await collections)[gameType];
+  }
+
+  // return the appropriate mongo collection used for the given game
+  private async _gameToCollection(game: GameType) {
```

Comment:
do we need this method if we are already maintaining `_gameEnumToCollection`? looks like we could just hydrate it with the collections initially in the constructor

Reply:
We have to await getting the collections, that was why we were having a hard time finding a way to map these to string or enums in the constructor

Cant have async in constructor